spyder/api/widgets.py

```diff
@@ -0,0 +1,1176 @@
+            sep.setSeparator(True)
+            sec_items.append((None, sep))
+
+        if sec_items:
+            sec_items.pop()
+
+        for (sec, item) in sec_items:
+            if isinstance(item, QAction):
+                add_method = super().addAction
+                insert_method = super().insertAction
+            else:
+                add_method = super().addWidget
+                insert_method = super().insertWidget
+
+            if self._corner_widget is not None:
+                insert_method(self._corner_separator, item)
+            else:
+                add_method(item)
+
+            if isinstance(item, QAction):
+                text_beside_icon = getattr(item, 'text_beside_icon', False)
+                if text_beside_icon:
+                    widget = self.widgetForAction(item)
+                    widget.setToolButtonStyle(Qt.ToolButtonTextBesideIcon)
+
+                if item.isCheckable():
+                    widget = self.widgetForAction(item)
+                    widget.setCheckable(True)
+
+    def create_toolbar_stretcher(self):
+        """
+        Create a stretcher widget to be used in a Qt toolbar.
+        """
+        stretcher = QWidget()
+        stretcher.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Expanding)
+        return stretcher
+
+    def _set_corner_widget(self, corner_widget):
+        """
+        Add the given corner widget to this toolbar.
+
+        A stretcher widget is added before the corner widget so that
+        its position is forced to the right side of the toolbar when the
+        toolbar is resized.
+        """
+        self._corner_widget = corner_widget
+        if corner_widget is not None:
+            stretcher = self.create_toolbar_stretcher()
+            self._corner_separator = super().addWidget(stretcher)
+            super().addWidget(self._corner_widget)
+        else:
+            self._corner_separator = None
+
+    def _setup_style(self):
+        """
+        Set the style of this toolbar with a stylesheet.
+        """
+        if is_dark_interface():
+            stylesheet = r"""
+                QToolBar QToolButton:!hover:!pressed {
+                    border-color: transparent;
+                }
+                QToolBar {
+                    border: 0px;
+                    background: rgb(25, 35, 45);
+                }
+                QToolButton {
+                    background-color: transparent;
+                }
+                QToolButton:checked {
+                    background-color: rgb(49, 64, 75);
+                }
+            """
+        else:
+            stylesheet = r"QToolBar {border: 0px;}"
+
+        self.setStyleSheet(textwrap.dedent(stylesheet))
+
+
+class PluginWidget(QWidget, SpyderWidgetMixin, SpyderToolBarMixin):
+    """
+    Spyder widget class.
+
+    This class handles non dockable widget.
+
+    Notes
+    -----
+    All Spyder non dockable plugins define a plugin widget that must subclass
+    this.
```

Comment:
Because Plugins are no longer widgets so they cannot mix implementations of plugins and widgets like before (which was quite a mess)

Even if a Plugin does not have a "Physical visible entity" it still needs to provide a Class to store the functionality and to be able to have a Class that is testable and usable without needing to use a Plugin.

A completion Plugin will live on this "Class" it HAS to be a QObject to be able to have signals, but if this holder Class is not also a QWidget it cannot parent child widgets like a status bar widget, so for instance in the case of completions.

An example:

```python
# spyder.plugins.kite.widgets.main_widget.py
class KiteCompletionClientWidget(PluginWidget, SpyderCompletionMixin):
      DEFAULT_OPTIONS = {}

      def __init__(self, ...)
            self.kite_status = KiteStatusWidget(self)

    # SpyderWidgetMixin API
    def setup(self, options=DEFAULT_OPTIONS):
            # Create actions, menus, etc

     def on_option_update(self, options=DEFAULT_OPTIONS):
            ...

     def update_actions(self, options=DEFAULT_OPTIONS):
            ...

      # The rest of the Completion API request, broadcast start etc....
     ....


# spyder.plugins.kite.plugin.py
class Kite(SpyderPluginV2):
     WIDGET_CLASS = KiteCompletionClientWidget
```